vcd/resource_vcd_ip_space.go

```diff
@@ -425,3 +478,97 @@ func setIpSpaceData(d *schema.ResourceData, ipSpace *types.IpSpace) error {
 
 	return nil
 }
+
+// getIpRangeIdFromFromPreviousState helps to find ip_range ID from previous state (because the current does not have it) and match it for current configuration
+func getIpRangeIdFromFromPreviousState(d *schema.ResourceData, startAddress, endAddress string) string {
+	ipRangesOld, _ := d.GetChange("ip_range")
+	ipRangesOldSchema := ipRangesOld.(*schema.Set)
+	ipRangesSlice := ipRangesOldSchema.List()
+
+	util.Logger.Printf("[TRACE] Looking for ID of 'ip_range' with start_address '%s' and end_address '%s'\n", startAddress, endAddress)
+
+	// Looping over ip_range definitions from state which contained all values and also stored ID. It looks for this ID with 2 priority levels:
+	// 1. An exact match with the same start and end IP addresses is found - return it immediately
+	// as this particular block has not changed at all for update operation
+	// 2. A partial match - if at least the start_address or end_address remained the same - cache
```

Comment:
Will give another think about it, but if we leave only an exact match - the risk is that then one cannot perform `update` to the range at all, and because of Terraform schema limitations we cannot flag that to user in diff (the only option we have is to Force whole resource creation, which is not what we want.

Reply:
So another thought about possible outcomes for the partial match case of IP range.

Because the "_partial match"_ matches the ID to either `start_address` or the `end_address` there is a risk that if one takes a range of 
```
start_address = 10.10.10.10
end_address  = 10.10.10.20
```


And in an update, he splits into two ranges where one matches the `start_address` and the other matches the `end_address` - we will get an update error ( duplicate IDs). The workaround would be to do that split in 2 update steps.
```
start_address = 10.10.10.10 <- matches previous start_address
end_address  = 10.10.10.14

start_address = 10.10.10.18
end_address  = 10.10.10.20 <- matches previous end_address
```

Another way is - we could be more conservative and **only match on `start_address`** (and not on `end_address`) for _"partial match"_, but the chances to update become a bit more limited. OTOH, Floating IP allocation (at least to what I noticed) usually starts from the start of the range. So if one has anything allocated, he should not be able to to change `start_address` during update and if he does not have it allocated - it should be fine to recreate that range (do the _update_ without specifying previous ID)

Reply:
> And in a single update, he splits into two ranges where one matches the start_address and the other matches the end_address - we will get an update error ( duplicate IDs).

Getting an error is fine. Most importantly, is there any chance for data corruption: when user sends one thing, but as a result gets back a different one?

Reply:
I have looked at this particular case manually. I had a range that had 3 floating IPs allocated for services.

So here is the diff I attempted:
```
     - ip_range {
          - end_address   = "11.11.11.111" -> null
          - id            = "cd10790c-5a65-4869-aeb2-b05568d28c01" -> null
          - start_address = "11.11.11.100" -> null
        }
      + ip_range {
          + end_address   = "11.11.11.108"
          + id            = (known after apply)
          + start_address = "11.11.11.100"
        }
      + ip_range {
          + end_address   = "11.11.11.111"
          + id            = (known after apply)
          + start_address = "11.11.11.110"
        }
```

Instead of one range:
```
start_address = "11.11.11.100"
end_address   = "11.11.11.111"
```

I put two ranges 
```
start_address = "11.11.11.100" <- matches the start_address of single range I had
end_address   = "11.11.11.108"


start_address = "11.11.11.110"
end_address   = "11.11.11.111" <- matches the end_address of single range I had
```

This error occurred during apply (Update):
```
vcd_ip_space.space1: Modifying... [id=urn:vcloud:ipSpace:ab75e441-d95c-4c43-9957-d5b8f7b10955]
╷
│ Error: error updating IP Space: error updating IP Space: error in HTTP PUT request: INTERNAL_SERVER_ERROR - [ fda98c3e-e04b-4210-b49c-2f2b1d04370f ] Duplicate key cd10790c-5a65-4869-aeb2-b05568d28c01 (attempted merging values 11.11.11.110 - 11.11.11.111 and 11.11.11.100 - 11.11.11.108)
│ 
│   with vcd_ip_space.space1,
│   on vcd.TestAccVcdIpSpaceIpAllocationstep2.tf line 45, in resource "vcd_ip_space" "space1":
│   45: resource "vcd_ip_space" "space1" {
│ 
```

In the UI - it remained unchanged - meaning that no changes occurred during this error. I tried running `terraform apply` 3 times - luckily it always presents the same diff and no changes occur in UI so this error looks to be addressable with no harm to configuration.

Then, I have attempted to do this "split" in two steps:
1. Shrunk existing range from `11.11.11.100 - 11.11.11.111` to `11.11.11.100 - 11.11.11.108` (this worked - update preserved correct ID and no impact to allocated IPs occurred)
2.  I added another range `11.11.11.110-11.11.11.111` which is just an addition of a new IP range now (it wouldn't have an old ID anyway)

Reply:
Interesting. What would happen with current matching mechanism if you did such a change:

```
start_address = "11.11.11.100"
end_address   = "11.11.11.111"
```

split into two with a caveat ->

```
start_address = "11.11.11.100"
end_address   = "11.11.11.108"

start_address = "11.11.11.111" <- the old end_address becomes a new start_address
end_address   = "11.11.11.118"
```

Reply:
In my mind I evaluated it to create new one for the second one - because the condition for partial match is this:

```go
if ipRangeStrings["start_address"] == startAddress || ipRangeStrings["end_address"] == endAddress {
```

That is an IP address and "position" (start or end) must match.

Here is the payload excerpt it created for `PUT`:
```json
...
"ipSpaceRanges": {
    "ipRanges": [
      {
        "startIpAddress": "11.11.11.111", <-- this did not match as previously it was `end_address`, no `id` field
        "endIpAddress": "11.11.11.118"
      },
      {
        "id": "8f7ed9d5-6779-4c4e-80af-b9ce14fad3db",
        "startIpAddress": "11.11.11.100",  <--- `start_address` before and after matched - `id` is present
        "endIpAddress": "11.11.11.108"
      }
    ],
    "defaultFloatingIpQuota": 0
  },
....
```


I also tried the reverse scenario - from:
```
start_address = "11.11.11.100"
end_address   = "11.11.11.108"

start_address = "11.11.11.111"
end_address   = "11.11.11.118"
```

Change to 
```
start_address = "11.11.11.100"
end_address   = "11.11.11.111"
```

It also worked well, picked the correct ID because it matched `start_address = "11.11.11.100"`. 

Both update scenarios worked given that that this range still had 3 floating IPs allocated - `11.11.11.100`, `11.11.11.101`, `11.11.11.102`

Reply:
That's great! So, it seems the only apparent issue is how we handle when the user does hit that error:

> │ Error: error updating IP Space: error updating IP Space: error in HTTP PUT request: INTERNAL_SERVER_ERROR - [ fda98c3e-e04b-4210-b49c-2f2b1d04370f ] Duplicate key cd10790c-5a65-4869-aeb2-b05568d28c01 (attempted merging values 11.11.11.110 - 11.11.11.111 and 11.11.11.100 - 11.11.11.108)

I mean, the above will be very confusing for the user.

Reply:
So - to reiterate. This can happen in such case:
Splitting a single range:
```
start_address = "11.11.11.100"
end_address   = "11.11.11.111"
```

Into two ranges, where one matches `start_address` and another matches `end_address`
```
start_address = "11.11.11.100" <- matches the start_address of single range I had
end_address   = "11.11.11.108"


start_address = "11.11.11.110"
end_address   = "11.11.11.111" <- matches the end_address of single range I had
```

I could catch the error (Duplicate key) and add our own message. Does such a message make sense:
> Splitting a single IP Range into two ranges with matching `start_address` and `end_address` is not supported in a single update operation. Please perform it in two steps.

Reply:
Made such an attempt https://github.com/vmware/terraform-provider-vcd/pull/1062/files#diff-134d08acac0e541b396ca1ac4225bbcda87e08ecbcb8772049dbc6026d52b79fR195-R212

Reply:
Such precise error is exactly what we needed! To close this out, I'm reading new code: is there any other scenario where "duplicate" message may be returned and we may misleading the user with the "split" hint? Maybe the "if" should be more deterministic and check that it's actually a "split" use case?

Reply:
I have attempted to abuse partial matches to see if there are any other errors that can happen. Here is a list of what I have tried:

* **IP Prefix** (it does partial match _only on `first_ip`_ field.
To replicate a problem, I tried to create:
```hcl
    prefix {
      first_ip      = "192.168.1.100"
      prefix_length = 30
      prefix_count  = 4
    }

    prefix { <- added on update
      first_ip      = "192.168.1.100" <- duplicate `first_ip`
      prefix_length = 28
      prefix_count  = 1
    }
```

```
 Error: error updating IP Space: error updating IP Space: error in HTTP PUT request: BAD_REQUEST - [ 15a6839f-f675-4454-9cfd-fab815c9010b ] IP Space TestAccVcdIpSpaceIpAllocation has prefix length 30 mismatch in the configuration.
│ 
│   with vcd_ip_space.space1,
│   on vcd.TestAccVcdIpSpaceIpAllocationstep2.tf line 45, in resource "vcd_ip_space" "space1":
│   45: resource "vcd_ip_space" "space1" {
```


* **IP Range** (it does partial matching _either on `start_address` or `end_address`_

This is our known and tested scenario:
Splitting a single range:
```
start_address = "11.11.11.100"
end_address   = "11.11.11.111"
```

Into two ranges, where one matches `start_address` and another matches `end_address`
```
start_address = "11.11.11.100" <- matches the start_address of single range I had
end_address   = "11.11.11.108"


start_address = "11.11.11.110"
end_address   = "11.11.11.111" <- matches the end_address of single range I had
```

The error currently looks like this:
```
vcd_ip_space.space1: Modifying... [id=urn:vcloud:ipSpace:99746875-6942-4054-8030-77af980d9445]
╷
│ Error: error updating IP Space: splitting a single IP Range into two ranges with matching 'start_address' and 'end_address' is not supported in a single update operation. Please perform it in two steps: error updating IP Space: error in HTTP PUT request: INTERNAL_SERVER_ERROR - [ d20e9f9d-c985-4ef1-a60c-8b7be29d859b ] Duplicate key 7af38863-1af7-4481-9ee4-79c7ba782296 (attempted merging values 11.11.11.100 - 11.11.11.107 and 11.11.11.108 - 11.11.11.111)
│ 
│   with vcd_ip_space.space1,
│   on vcd.TestAccVcdIpSpaceIpAllocationstep2.tf line 45, in resource "vcd_ip_space" "space1":
│   45: resource "vcd_ip_space" "space1" {
```

* **Another IP Range  scenario** is adding second IP range with _duplicate `start_address` or `end_address`_ 

```hcl
  ip_range {
    start_address = "11.11.11.100"
    end_address   = "11.11.11.111"
  }

  ip_range { <- added during update
    start_address = "11.11.11.100"
    end_address   = "11.11.11.108"
  }
```

The error in both cases looks like this:
```
vcd_ip_space.space1: Modifying... [id=urn:vcloud:ipSpace:99746875-6942-4054-8030-77af980d9445]
╷
│ Error: error updating IP Space: error updating IP Space: error in HTTP PUT request: BAD_REQUEST - [ 4a62d80f-2d01-4270-b3d6-44ba13134b9e ] The following IP ranges contain an overlap: 11.11.11.100 - 11.11.11.111 and 11.11.11.100 - 11.11.11.108.
│ 
│   with vcd_ip_space.space1,
│   on vcd.TestAccVcdIpSpaceIpAllocationstep2.tf line 45, in resource "vcd_ip_space" "space1":
│   45: resource "vcd_ip_space" "space1" {
│ 
╵
```

Reply:
This is great, thank you! I'm resolving this conversation. Also, approving the PR :)